### PR TITLE
VRFV2 Wrapper: Contracts

### PR DESCRIPTION
**Pull request by Nic0S**

### CODE REVIEW

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+import "../interfaces/VRFCoordinatorV2Interface.sol";
+import "../vendor/SafeMathChainlink.sol";
+import "./VRFV2WrapperInterface.sol";
+
+/**
+ * @notice A wrapper for VRFCoordinatorV2 that provides an interface better suited to one-off
+ * @notice requests for randomness.
+ */
+contract VRFV2Wrapper is ConfirmedOwner, TypeAndVersionInterface, VRFConsumerBaseV2, VRFV2WrapperInterface {
+  using SafeMathChainlink for uint256;
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:20]: I wonder if this is necessary for Solidity 0.8? I thought that it comes by default with 0.8 (https://soliditydeveloper.com/solidity-0.8)

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 2]
```diff
@@ -0,0 +1,260 @@
...
+
+  LinkTokenInterface public immutable LINK;
+  AggregatorV3Interface public immutable LINK_ETH_FEED;
+  ExtendedVRFCoordinatorV2Interface public immutable COORDINATOR;
+  uint64 public immutable SUBSCRIPTION_ID;
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:25]: Do we foresee a situation where we'd might have to mutate this? I can't think of any off the top of my head, besides somehow the subscription owner being compromised (though if the owner is a Gnosis safe, should be OK?)

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 3]
```diff
@@ -0,0 +1,260 @@
...
+
+  // lastRequestId is the request ID of the most recent VRF V2 request made by this wrapper. This
+  // should only be relied on within the same transaction the request was made.
+  uint256 public override lastRequestId;
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:29]: I'm curious, what does `override` mean in the context of a storage field?

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 4]
```diff
@@ -0,0 +1,260 @@
...
+  // requests cannot be made.
+  bool public s_configured;
+
+  // s_disabled disables the contract when true. When disabled, new VRF requests cannot be made
+  // but existing ones can still be fulfilled.
+  bool public s_disabled;
+
+  // s_fallbackWeiPerUnitLink is the backup LINK exchange rate used when the LINK/NATIVE feed is
+  // stale.
+  int256 private s_fallbackWeiPerUnitLink;
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:43]: Can't pull this from the coordinator directly? Similar question for the other configuration fields below.

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 5]
```diff
@@ -0,0 +1,260 @@
...
+      s_keyHash,
+      s_maxNumWords
+    );
+  }
+
+  function calculatePrice(uint32 callbackGasLimit) public view override onlyConfiguredNotDisabled returns (uint256) {
+    return calculatePriceInternal(callbackGasLimit, tx.gasprice);
+  }
+
+  function estimatePrice(uint32 _callbackGasLimit, uint256 _requestGasPrice)
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:153]: Nit: Might be worth being super verbose here and calling the gas price parameter `_requestGasPriceWei` (since it's common that it gets specified in gwei instead).

File: core/scripts/vrfv2/testnet/main.go [hunk 6]
```diff
@@ -21,6 +21,8 @@ import (
 	"github.com/shopspring/decimal"
 
 	"github.com/smartcontractkit/chainlink/core/assets"
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/vrfv2_wrapper"
+	"github.com/smartcontractkit/chainlink/core/internal/gethwrappers/generated/vrfv2_wrapper_consumer_example"
```
Comment (makramkd) [inline core/scripts/vrfv2/testnet/main.go:25]: Small nit: I think these imports are in the wrong place (should be grouped with the other gethwrappers stuff further down).

**Review by makramkd — COMMENTED** [review]
O for Owesome. Very minor stuff.

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 7]
```diff
@@ -0,0 +1,260 @@
...
+      weiPerUnitLink = s_fallbackWeiPerUnitLink;
+    }
+    return weiPerUnitLink;
+  }
+
+  function onTokenTransfer(
+    address _sender,
+    uint256 _amount,
+    bytes calldata _data
+  ) external onlyConfiguredNotDisabled {
```
Comment (vreff) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:191]: Very cool to leverage transferAndCall as a direct request mechanism.

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 8]
```diff
@@ -0,0 +1,260 @@
...
+    require(msg.sender == address(LINK), "only callable from LINK");
+
+    (uint32 callbackGasLimit, uint16 requestConfirmations, uint32 numWords) = abi.decode(
+      _data,
+      (uint32, uint16, uint32)
+    );
+    uint256 price = calculatePrice(callbackGasLimit);
+    require(_amount >= price, "fee too low");
```
Comment (vreff) [inline contracts/src/v0.8/dev/VRFV2Wrapper.sol:199]: More of a question here, is there a mandate now to use custom errors, or are we still ok to use `require`? I like require better still for readability & since tooling is not good for custom errors yet, but just curious.

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol (hunk 1 above)
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol]: TIL! removed

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol (hunk 2 above)
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol:21]: Yup! That's why I think it's safe to make this immutable, it should be 100% impossible to change anything about the subscription since the wrapper doesn't expose any way to do that

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol (hunk 4 above)
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol:39]: Right yeah, if coordinator config changes we need to make a call to update the config here too. Could get the config each time but adds more gas cost, I'm open to either. In the end though, this should only affect profitability so I'd rather take the gas savings for users

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+      s_fulfillmentFlatFeeLinkPPM,
+      s_gasAfterPaymentCalculation,
+      s_wrapperPremiumPercentage,
+      s_wrapperPremiumPercentage,
+      s_keyHash,
+      s_maxNumWords
+    );
+  }
+
+  function calculatePrice(uint32 callbackGasLimit) public view override onlyConfiguredNotDisabled returns (uint256) {
```
[the remark this answers is not in the document]
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol]: Done

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol (hunk 5 above)
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol]: Done

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+  }
+
+  function estimatePrice(uint32 _callbackGasLimit, uint256 _requestGasPrice)
+    external
+    view
+    onlyConfiguredNotDisabled
+    returns (uint256)
+  {
+    return calculatePriceInternal(_callbackGasLimit, _requestGasPrice);
+  }
```
[the remark this answers is not in the document]
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol]: Right yeah, I thought it was worth adding so you could call from a block explorer. I think in an eth_call, you could actually call calculatePrice as long as you set gas price in the simulation but haven't tested.

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol (hunk 7 above)
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol:232]: Yeah we do this in a lot of spots, I like the pattern. VRF V1 does this too

File: contracts/src/v0.8/dev/VRFV2Wrapper.sol (hunk 8 above)
Comment (Nic0S) [reply contracts/src/v0.8/dev/VRFV2Wrapper.sol:240]: Yeah I went this way because of the tooling, we still don't know how to parse out the custom errors haha

File: core/scripts/vrfv2/testnet/main.go (hunk 6 above)
Comment (Nic0S) [reply core/scripts/vrfv2/testnet/main.go]: fixed!

**Review by makramkd — DISMISSED** [review]
:shipit:

File: contracts/src/v0.8/dev/VRFV2WrapperConsumerBase.sol [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+/** *******************************************************************************
+ * @notice Interface for contracts using VRF randomness through the VRF V2 wrapper
+ * ********************************************************************************
+ * @dev PURPOSE
+ *
+ * @dev Create VRF V2 requests without the need for subscription management. Rather than creating
+ * @dev and funding a VRF V2 subscription, a user can use this wrapper to create one off requests,
+ * @dev paying up front rather than at fulfillment. After fulfillment, a refund is issued for
+ * @dev any unused gas in the callback (if any).
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2WrapperConsumerBase.sol:15]: Should remove this comment on the refund, since we're not doing that.

**Review by makramkd — COMMENTED** [review]
Just thought of this now, but I think it would be useful to have an integration test, or maybe a couple of we have a bunch of test cases in mind (in core/services/vrf/integration_v2_test.go) that deploys this wrapper contract to the simulated blockchain and makes some requests there.

File: contracts/src/v0.8/dev/VRFV2WrapperConsumerBase.sol [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+ * @dev USAGE
+ *
+ * @dev Calling contracts must inherit from VRFV2WrapperConsumerBase. The consumer must be funded
+ * @dev with enough LINK to make the request, otherwise requests will revert. To request randomness,
+ * @dev call the 'requestRandomness' function with the desired VRF parameters. This function handles
+ * @dev paying for the request based on the current pricing.
+ *
+ * @dev Consumers must implement the fullfillRandomWords function, which will be called during
+ * @dev fulfillment with the randomness result. The consumer will be issued a LINK refund for gas
+ * @dev that was unused during the callback (if any).
```
Comment (makramkd) [inline contracts/src/v0.8/dev/VRFV2WrapperConsumerBase.sol:30]: Can remove this refund wording, since we're no longer doing that.

File: core/services/vrf/mocks/aggregator_v3_interface.go [hunk 13]
```diff
@@ -1,4 +1,4 @@
-// Code generated by mockery v2.13.0-beta.1. DO NOT EDIT.
+// Code generated by mockery v2.12.2. DO NOT EDIT.
```
Comment (makramkd) [inline core/services/vrf/mocks/aggregator_v3_interface.go:1]: I think you should probably revert the changes on these files, since you're using an older version of mockery.

File: core/scripts/vrfv2/testnet/main.go [hunk 14]
```diff
@@ -832,6 +834,108 @@ func main() {
...
+
+		consumer, err := vrfv2_wrapper_consumer_example.NewVRFV2WrapperConsumerExample(
+			common.HexToAddress(*consumerAddress), e.Ec)
+		helpers.PanicErr(err)
+
+		tx, err := consumer.MakeRequest(e.Owner, uint32(*cbGasLimit), uint16(*confirmations), uint32(*numWords))
+		helpers.PanicErr(err)
+		helpers.ConfirmTXMined(context.Background(), e.Ec, tx, e.ChainID)
+	case "wrapper-universe-deploy":
+		cmd := flag.NewFlagSet("wrapper-universe-deploy", flag.ExitOnError)
```
Comment (makramkd) [inline core/scripts/vrfv2/testnet/main.go:889]: Can extract this into it's own method (like `deployWrapperUniverse`) similar to `deployUniverse` for VRF v2.

**Review by makramkd — APPROVED** [review]
